### PR TITLE
Save players

### PR DESCRIPTION
**Pull request by dex-89**
**Player save in case of server crash**

### CODE REVIEW

File: data/lib/core/storages.lua [hunk 1]
```diff
@@ -135,6 +135,7 @@ Storage = {
 	LemonCupcake = 30052,
 	BlueberryCupcake = 30053,
 	PremiumAccount = 998899,
+	SavePlayer = 1111,
```
Comment (dudantas) [inline data/lib/core/storages.lua]: Can you use the existing range of storages? 30053, 30054, 30055 ...

Comment (dex-89) [reply data/lib/core/storages.lua]: No problem. Fixed

**Review by dudantas — APPROVED** [review]

File: data/lib/core/storages.lua (hunk 1 above)
Comment (dudantas) [reply data/lib/core/storages.lua]: Thanks

**Review by mattyx14 — APPROVED** [review]

**Comment by dex-89** [comment]
I try make anti-rollback system..

**Comment by DeCarvalhoJr** [comment]
Adding it was making the console return this failure line
`Failure: [Event::loadCallback] scriptInterface == nullptr. scriptid = 1604`

Server was able to start but i didnt notice it working, tried to reduce saev time in order to be able to have it working, yet, nothing  new.

Removing it made the server start normally.

**Comment by dex-89** [comment]
> Adding it was making the console return this failure line
> `Failure: [Event::loadCallback] scriptInterface == nullptr. scriptid = 1604`
> 
> Server was able to start but i didnt notice it working, tried to reduce saev time in order to be able to have it working, yet, nothing new.
> 
> Removing it made the server start normally.

Thanks for your test. I will try to correct it so that there are no mistakes

**Comment by dex-89** [comment]
> Adding it was making the console return this failure line
> `Failure: [Event::loadCallback] scriptInterface == nullptr. scriptid = 1604`
> 
> Server was able to start but i didnt notice it working, tried to reduce saev time in order to be able to have it working, yet, nothing new.
> 
> Removing it made the server start normally.

Can you test it again? After fix?

**Review by mattyx14 — APPROVED** [review]

**Comment by DeCarvalhoJr** [comment]
Hi, tested again and once again got an error.
it staarted to spam the error every 15 seconds after having my char logged in

The following error occurs, as stated, every 15 seconds
```
2020-12-16 08:05:39 -  Lua Script Error: [Main Interface]
2020-12-16 08:05:39 -  in a timer event called from:
2020-12-16 08:05:39 -  (Unknown scriptfile)
2020-12-16 08:05:39 -  .../otbr/data/scripts/creaturescripts/others/saveplayer.lua:9: attempt to call global 'doPlayerSave' (a nil value)
2020-12-16 08:05:39 -  stack traceback:
2020-12-16 08:05:39 -   [C]: in function 'doPlayerSave'
2020-12-16 08:05:39 -   .../otbr/data/scripts/creaturescripts/others/saveplayer.lua:9: in function 'doSavePlayerAndHouse'
2020-12-16 08:05:39 -   .../otbr/data/scripts/creaturescripts/others/saveplayer.lua:17: in function <.../otbr/data/scripts/creaturescripts/others/saveplayer.lua:14>
```

i did change two things only from saveplayer.lua and didnt touch anything aside it.
```
    saveInterval = 1 * 15,
    minSaveInterval = 1 * 15,
    maxSaveInterval = 2 * 30,
```

Some information, just in case.
```
Chassis: vm
Virtualization: kvm
Operating System: Debian GNU/Linux 10 (buster)
Kernel: Linux 4.19.0-10-cloud-amd64
Architecture: x86-64
```

Had to remove in order to have my server running normally.
Lemme know about the fix so i can test it again, sorry for being late on testing it.
Cheers

//editing
after changing back src files and everything that i had to, i forgot to remove the saveplayer.lua.. did recompile to be sure it would generate a new file to start the server and... weirdly.. the same error started to occur...
Twas like nothing changed.. I mean, the error could be happening due changes in src nothing effetcing anything at all.
I have the latest version of OTBR, with minor changes made by me, nothing to do with this part, just vip system and npc selling spells..

Anyway, thought it would be good to add here.

Cheers once again

**Comment by dex-89** [comment]
> Hi, tested again and once again got an error.
> it staarted to spam the error every 15 seconds after having my char logged in
> 
> The following error occurs, as stated, every 15 seconds
> 
> ```
> 2020-12-16 08:05:39 -  Lua Script Error: [Main Interface]
> 2020-12-16 08:05:39 -  in a timer event called from:
> 2020-12-16 08:05:39 -  (Unknown scriptfile)
> 2020-12-16 08:05:39 -  .../otbr/data/scripts/creaturescripts/others/saveplayer.lua:9: attempt to call global 'doPlayerSave' (a nil value)
> 2020-12-16 08:05:39 -  stack traceback:
> 2020-12-16 08:05:39 -   [C]: in function 'doPlayerSave'
> 2020-12-16 08:05:39 -   .../otbr/data/scripts/creaturescripts/others/saveplayer.lua:9: in function 'doSavePlayerAndHouse'
> 2020-12-16 08:05:39 -   .../otbr/data/scripts/creaturescripts/others/saveplayer.lua:17: in function <.../otbr/data/scripts/creaturescripts/others/saveplayer.lua:14>
> ```
> 
> i did change two things only from saveplayer.lua and didnt touch anything aside it.
> 
> ```
>     saveInterval = 1 * 15,
>     minSaveInterval = 1 * 15,
>     maxSaveInterval = 2 * 30,
> ```
> 
> Some information, just in case.
> 
> ```
> Chassis: vm
> Virtualization: kvm
> Operating System: Debian GNU/Linux 10 (buster)
> Kernel: Linux 4.19.0-10-cloud-amd64
> Architecture: x86-64
> ```
> 
> Had to remove in order to have my server running normally.
> Lemme know about the fix so i can test it again, sorry for being late on testing it.
> Cheers
> 
> //editing
> after changing back src files and everything that i had to, i forgot to remove the saveplayer.lua.. did recompile to be sure it would generate a new file to start the server and... weirdly.. the same error started to occur...
> Twas like nothing changed.. I mean, the error could be happening due changes in src nothing effetcing anything at all.
> I have the latest version of OTBR, with minor changes made by me, nothing to do with this part, just vip system and npc selling spells..
> 
> Anyway, thought it would be good to add here.
> 
> Cheers once again

Damn. Github didn't save one change. Change doPlayerSave(cid) to Player(cid):save()

**Review by DeCarvalhoJr — CHANGES_REQUESTED** [review]
Hey, i performed the test.. 

So, `Change doPlayerSave(cid) to Player(cid):save()` made thru it..
You should add the change to the file, considering it isnt delivering any error by now.

im about to ask if you could add a way to it send a console msg before saving, so we can follow it. 
Again, made the change
```
    saveInterval = 1 * 15,
    minSaveInterval = 1 * 15,
    maxSaveInterval = 2 * 30,
```
hoping it could gimme at least a console [log] and nothing. Im logged in for more than 20 minutes already.

Cheers

**Comment by DeCarvalhoJr** [comment]
Hi, was looking into your changes and i have a doubt about the consoleLog
what if adm set log to true and he has, lets think about 100 players online, wouldnt it spam the name of every player? if so, wouldnt it cause some loss on performance?
I thought about a general log, just like 'players have been saved' or so, then we would be able to know it was working.
im asking cuz im unsure if im wrong or right about this.

Cheers

**Comment by dex-89** [comment]
> Hi, was looking into your changes and i have a doubt about the consoleLog
> what if adm set log to true and he has, lets think about 100 players online, wouldnt it spam the name of every player? if so, wouldnt it cause some loss on performance?
> I thought about a general log, just like 'players have been saved' or so, then we would be able to know it was working.
> im asking cuz im unsure if im wrong or right about this.
> 
> Cheers

Global save causes more lags. Log has been added only at your request. In this version, player save starts when logging in and ends when logging out.

**Review by DeCarvalhoJr — APPROVED** [review]
Aparently eveything is working.
Still, i maintain my concern about a server having 30/40/50+ players and the system saving it every 60sec, ofc, we can configure it, anyway, wont it cause a lag even if saving every 1hr, something like 100 players?
Its just a concern im leaving here..

Im gonna mark as 'approve' cuz i found no error and console didnt give an error.

```
Hell Train has logged in.
Player Hell Train has been saved
Player Hell Train has been saved
Player Hell Train has been saved
Player Hell Train has been saved
Player Hell Train has been saved
Player Hell Train has been saved
Player Hell Train has been saved
Player Hell Train has been saved
Player Hell Train has been saved
Player Hell Train has been saved
Player Hell Train has been saved
```
```
local config = { --Times are in seconds
	saveInterval = 1 * 60,
	minSaveInterval = 1 * 30,
	maxSaveInterval = 2 * 30,
	storage = Storage.SavePlayer,
	consoleLog = true -- set as true to display logs in console
}
```

**Comment by dex-89** [comment]
Too many mistakes

**Pull request closed without merging**